### PR TITLE
Merge phdr and shdr and enhance plt detection

### PR DESCRIPTION
**Pull request by radare**

### CODE REVIEW

File: libr/bin/format/elf/elf.c [hunk 1]
```diff
@@ -1250,6 +1253,9 @@ static ut64 get_import_addr(ELFOBJ *bin, int sym) {
 			case EM_ARM:
 			case EM_AARCH64:
 				plt_addr = Elf_(r_bin_elf_get_section_addr) (bin, ".plt");
+				if (plt_addr == 0) {
+					plt_addr = Elf_(r_bin_elf_get_section_addr_end) (bin, ".rela.plt");
```
Comment (SrimantaBarua) [inline libr/bin/format/elf/elf.c]: Not always correct

**Comment by radare** [comment]
I know but its a workaround for the test i was addressing

**Comment by radare** [comment]
Thats why its not merged. This and because sometimes the sections appear in different order so it results in wrong listing that breaks tests

**Comment by SrimantaBarua** [comment]
```
#if R_BIN_ELF64
		dyn[i].d_tag = READ64 (sdyn, j)
		dyn[i].d_un.d_ptr = READ64 (sdyn, j)
#else
		dyn[i].d_tag = READ32 (sdyn, j)
		dyn[i].d_un.d_ptr = READ32 (sdyn, j)
#endif
```
Won't things like this break when you open a 32-bit ELF on a 64-bit system?

**Comment by radare** [comment]
Why? This code is duplicated magically for 32 and 64 bits. Been there for years and working fine

**Comment by SrimantaBarua** [comment]
Oo okay I get it now.

**Comment by SrimantaBarua** [comment]
Rebased, added a few fixes. It is now at https://github.com/SrimantaBarua/radare2/tree/elf-merge-shdr-phdr

**Comment by radare** [comment]
What about sending a PR? Right now if u rebase we will not see the real brokeness but will be good to review your changes asap and get this fix merged

**Comment by SrimantaBarua** [comment]
This seems to not be very straightforward. Fixing `t/cmd_open` seems to be breaking the third test in `t.formats/elf/symbols`. So I'm reluctant to send a PR because I haven't made much progress :/

The basic changes which work (github forces me to rename it to .txt) -
[elf_patch_1.txt](https://github.com/radare/radare2/files/1247843/elf_patch_1.txt)

Issues -
1. `ut64 get_import_addr()` sometimes returns -1 and at other times return `UT32_MAX` on failure. The code in `get_symbols_from_phdr` sets `toffset` as 0 if `get_import_addr` returns -1, and then resolves the symbol. There is no case for `UT32_MAX`. That's why the test in `t.formats/elf/symbols` fails since 30 symbols are not reported.
When I add checking for `UT32_MAX` however, `t/cmd_open` fails because you get extra symbols reported.
2. `nsym` in `get_symbols_from_phdr` is set to a very high value (`(bin->size - addr_sym_table) / sym_size`). Limiting this to the number of symbols in `.dynsym` fixes `t/cmd_open` after changes in (1) above, but in `t.formats/elf/symbols` again 30 symbols are not reported.
3. In `get_import_addr`, instead of selecting one section like `rel.plt`, iterating over all relocation sections reports all the symbols in `t.formats/elf/symbols`, but they are at incorrect (non-zero) offsets. Plus it reports two extra symbols in `t/cmd_open`.

**Comment by radare** [comment]
ok ill rebase the pr with your patch

**Comment by radare** [comment]
uhm, this patch cant be applied in any commit.

**Comment by radare** [comment]
i have rebased my branch on top of master, still not green, but getting closer

**Comment by radare** [comment]
ping?

**Comment by alvarofe** [comment]
rebase please, the section thing get things wrong

```
idx=10 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid10
idx=11 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid11
idx=12 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid12
idx=13 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid13
idx=14 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid14
idx=15 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid15
idx=16 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid16
idx=17 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid17
idx=18 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid18
idx=19 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid19
idx=20 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid20
idx=21 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid21
idx=22 vaddr=0x00000000 paddr=0x00000000 sz=0 vsz=0 perm=----- name=invalid22
```

**Comment by radare** [comment]
cc @ret2libc

**Comment by ret2libc** [comment]
Before commenting on this I'd like to have a bit more of context. What is this trying to solve? Do we have tests of a bad behaviour in the current master branch? How does this improve the situation?

**Comment by radare** [comment]
good question, dont remember which was the sample bin we were using @alvarofe @SrimantaBarua ?

**Comment by radare** [comment]
maybe this one?

screen shot 2018-07-11 at 17 50 19

**Comment by radare** [comment]
or the one posted by the LIEF guys

**Comment by ret2libc** [comment]
Without a clear bug and reproducer, it's hard to review this one

**Comment by radare** [comment]
I think this was the issue found by the lief or bn guys that was showing puts in the imports instead of system. There was a poc and a tweet but i cant find it now. Maybe someone not in vacations can use a non-phone to google for it

**Comment by XVilka** [comment]
@radare you are speaking about textile, yes. It is because of how Linux kernel handles ELF (not following specification by the way).

**Comment by radare** [comment]
Thats why we have asm.os to select the kernel. We have also a wip implementation for an os-independent runtime linker for static analysis. Will be good to take care of it at some point

File: libr/bin/format/elf/elf.c (hunk 1 above)
Comment (MaskRay) [reply libr/bin/format/elf/elf.c]: ？？？？？？？？don't understand to use `.rela.plt`

Comment (MaskRay) [reply libr/bin/format/elf/elf.c]: I think phdr and shdr should be two separate views.

If you want to use phdr, use `DT_JMPREL` (offset) `DT_PLTRELSZ` (size) `DT_PLTREL` (element size).
`DT_REL` and `DT_RELA`

`musl/ldso/dynlink.c`

**Comment by ret2libc** [comment]
Closing. It was stale for so long and it's not even clear if it's a good change so far.

**Pull request closed without merging**